File: docs/sso-templates/typescript/casdoor.ts
/**
 * Casdoor JWT verification module — TypeScript
 *
 * 目标：工单系统 (gongdan) / 运营中心 / 任何 TS/Node 服务复制此文件即可接入。
 *
 * 依赖：
 *   npm i jose axios
 *
 * 环境变量：
 *   CASDOOR_ENDPOINT         https://casdoor.ashyglacier-8207efd2.eastasia.azurecontainerapps.io
 *   CASDOOR_ORG              xingyun
 *   CASDOOR_APP_NAME         ticket-app        # 每个系统不同
 *   CASDOOR_CLIENT_ID        xxx               # 从 Casdoor 后台拷
 *   CASDOOR_CLIENT_SECRET    xxx
 *   CASDOOR_REDIRECT_URI     https://<app-fqdn>/api/auth/callback
 *   CASDOOR_CERT             (可选) 直接贴 PEM 公钥
 */

import { importSPKI, jwtVerify, type JWTPayload } from 'jose';
import axios from 'axios';

export interface CasdoorConfig {
  endpoint: string;
  org: string;
  appName: string;
  clientId: string;
  clientSecret: string;
  redirectUri: string;
  certPem?: string;
}

export interface CasdoorClaims extends JWTPayload {
  name?: string;
  displayName?: string;
  email?: string;
  owner?: string;
  roles?: Array<string | { name: string }>;
  groups?: string[];
}

export interface CurrentUser {
  sub: string;
  name: string;
  email: string;
  owner: string;
  roles: string[];
  raw: CasdoorClaims;
}

let cachedPem: string | null = null;

async function loadPublicKeyPem(cfg: CasdoorConfig): Promise<string> {
  if (cachedPem) return cachedPem;
  if (cfg.certPem?.includes('BEGIN')) {
    cachedPem = cfg.certPem.replace(/\\n/g, '\n');
    return cachedPem;
  }
  const url = `${cfg.endpoint.replace(/\/$/, '')}/api/get-cert`;
  const res = await axios.get(url, { params: { id: `${cfg.org}/${cfg.appName}` } });
  const cert = res.data?.data?.certificate || res.data?.certificate;
  if (!cert) throw new Error(`empty cert from Casdoor: ${JSON.stringify(res.data)}`);
  cachedPem = cert;
  return cert;
}

export async function verifyToken(token: string, cfg: CasdoorConfig): Promise<CurrentUser> {
  const pem = await loadPublicKeyPem(cfg);
  const key = await importSPKI(pem, 'RS256');

  const { payload } = await jwtVerify(token, key, {
    algorithms: ['RS256'],
    audience: cfg.clientId,
    issuer: cfg.endpoint.replace(/\/$/, ''),
  });

  const claims = payload as CasdoorClaims;
  const rolesField = claims.roles || [];
  const roles = rolesField.map((r) => (typeof r === 'string' ? r : r.name)).filter(Boolean);

  return {
    sub: String(claims.sub || ''),
    name: claims.name || claims.displayName || '',
    email: claims.email || '',
    owner: claims.owner || '',
    roles,
    raw: claims,
  };
}

export async function exchangeCodeForToken(
  code: string,
  cfg: CasdoorConfig
): Promise<Record<string, unknown>> {
  const url = `${cfg.endpoint.replace(/\/$/, '')}/api/login/oauth/access_token`;
  const body = new URLSearchParams({
    grant_type: 'authorization_code',
    code,
    client_id: cfg.clientId,
    client_secret: cfg.clientSecret,
    redirect_uri: cfg.redirectUri,
  });
  const res = await axios.post(url, body.toString(), {
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  });
  return res.data;
}

export function authorizeUrl(cfg: CasdoorConfig, state = 'sso'): string {
  const qs = new URLSearchParams({
    client_id: cfg.clientId,
    response_type: 'code',
    redirect_uri: cfg.redirectUri,
    scope: 'read',
    state,
  });
  return `${cfg.endpoint.replace(/\/$/, '')}/login/oauth/authorize?${qs}`;
}

File: docs/sso-templates/go/casdoor.go
// Casdoor JWT verification — Go
//
// go get github.com/golang-jwt/jwt/v5
//
// Use for cloud management system (云管) or any non-Python/TS service.

package auth

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type Config struct {
	Endpoint     string
	Org          string
	AppName      string
	ClientID     string
	ClientSecret string
	RedirectURI  string
	CertPEM      string // optional, else fetched from Casdoor
}

type Claims struct {
	Name  string   `json:"name"`
	Email string   `json:"email"`
	Owner string   `json:"owner"`
	Roles []string `json:"roles"` // Casdoor may send []string OR []Role{}; normalize at marshal time
	jwt.RegisteredClaims
}

var (
	pemCache string
	pemOnce  sync.Once
	pemErr   error
)

func loadPem(cfg Config) (string, error) {
	pemOnce.Do(func() {
		if strings.Contains(cfg.CertPEM, "BEGIN") {
			pemCache = strings.ReplaceAll(cfg.CertPEM, "\\n", "\n")
			return
		}
		u := strings.TrimRight(cfg.Endpoint, "/") + "/api/get-cert?id=" + url.QueryEscape(cfg.Org+"/"+cfg.AppName)
		client := &http.Client{Timeout: 10 * time.Second}
		resp, err := client.Get(u)
		if err != nil {
			pemErr = err
			return
		}
		defer resp.Body.Close()
		var data struct {
			Data struct {
				Certificate string `json:"certificate"`
			} `json:"data"`
			Certificate string `json:"certificate"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
			pemErr = err
			return
		}
		cert := data.Data.Certificate
		if cert == "" {
			cert = data.Certificate
		}
		if cert == "" {
			pemErr = errors.New("empty cert from casdoor")
			return
		}
		pemCache = cert
	})
	return pemCache, pemErr
}

// VerifyToken validates a Casdoor-issued JWT and returns typed claims.
func VerifyToken(tokenStr string, cfg Config) (*Claims, error) {
	pem, err := loadPem(cfg)
	if err != nil {
		return nil, fmt.Errorf("load pem: %w", err)
	}
	key, err := jwt.ParseRSAPublicKeyFromPEM([]byte(pem))
	if err != nil {
		return nil, fmt.Errorf("parse pem: %w", err)
	}

	tok, err := jwt.ParseWithClaims(tokenStr, &Claims{}, func(t *jwt.Token) (interface{}, error) {
		if t.Method.Alg() != "RS256" {
			return nil, fmt.Errorf("unexpected alg: %v", t.Header["alg"])
		}
		return key, nil
	},
		jwt.WithAudience(cfg.ClientID),
		jwt.WithIssuer(strings.TrimRight(cfg.Endpoint, "/")),
	)
	if err != nil {
		return nil, err
	}
	if !tok.Valid {
		return nil, errors.New("invalid token")
	}
	c, _ := tok.Claims.(*Claims)
	return c, nil
}

// HasRole reports whether the claims include any of the required roles.
func (c *Claims) HasRole(roles ...string) bool {
	for _, r := range c.Roles {
		for _, want := range roles {
			if r == want {
				return true
			}
		}
	}
	return false
}
